### PR TITLE
Implements WatsonX model provider

### PR DESCRIPTION
**Pull request by rishacha**
## Description

Adds support for WatsonX as a model provider

## Checklist

- [x] The base branch of this PR is `dev`, rather than `main`
- [x] The relevant docs, if any, have been updated or created

Implements #1823 

## Screenshots

[ "For visual changes, include screenshots." ]

## Testing

[ For new or modified features, provide testing instructions. ]

### CODE REVIEW

File: core/llm/llms/Watsonx.ts
Comment (rishacha) [inline core/llm/llms/Watsonx.ts:1]: Keep this

File: core/llm/llms/index.ts
Comment (rishacha) [inline core/llm/llms/index.ts:1]: Keep this

File: core/llm/index.ts
Comment (rishacha) [inline core/llm/index.ts:1]: Keep this

File: core/config/types.ts
Comment (rishacha) [inline core/config/types.ts:1]: Keep this

File: core/index.d.ts
Comment (rishacha) [inline core/index.d.ts:1]: Keep this

File: docs/docs/reference/Model Providers/watsonx.md
Comment (rishacha) [inline docs/docs/reference/Model Providers/watsonx.md:1]: Keep this

File: docs/docs/setup/model-providers.md
Comment (rishacha) [inline docs/docs/setup/model-providers.md:1]: Keep this

File: docs/static/schemas/config.json
Comment (rishacha) [inline docs/static/schemas/config.json:1]: Keep this

File: extensions/intellij/src/main/resources/config_schema.json
Comment (rishacha) [inline extensions/intellij/src/main/resources/config_schema.json:1]: Keep this

File: extensions/vscode/config_schema.json
Comment (rishacha) [inline extensions/vscode/config_schema.json:1]: Keep this

File: extensions/vscode/continue_rc_schema.json
Comment (rishacha) [inline extensions/vscode/continue_rc_schema.json:1]: Keep this

File: gui/public/logos/WatsonX.png
Comment (rishacha) [inline gui/public/logos/WatsonX.png:1]: Keep this

File: gui/src/pages/AddNewModel/configs/models.ts
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/models.ts:1]: Keep this

File: gui/src/pages/AddNewModel/configs/providers.ts
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts:1]: Keep this

File: gui/src/pages/AddNewModel/ConfigureProvider.tsx
Comment (rishacha) [inline gui/src/pages/AddNewModel/ConfigureProvider.tsx:1]: Keep this

**Review by rishacha — COMMENTED** [review]
Please remove the unnecessary files

File: docs/static/schemas/config.json
Comment (B-Rajpal) [reply docs/static/schemas/config.json:1]: not required

File: core/index.d.ts [hunk 1]
```diff
@@ -590,7 +598,8 @@ type ModelProvider =
   | "deepseek"
   | "azure"
   | "openai-aiohttp"
-  | "msty";
+  | "msty"
+  | "WatsonX";
```
Comment (rishacha) [inline core/index.d.ts]: make it small case

File: core/index.d.ts [hunk 2]
```diff
@@ -339,6 +339,14 @@ export interface LLMOptions {
...
   region?: string;
   projectId?: string;
+
+  // WatsonX options
+  WatsonxUrl?: string;
+  WatsonxApiKey?: string;
+  WatsonxZenApiKeyBase64?:string// Required if using watsonx software with ZenApiKey auth
+  WatsonxUsername?:string;
+  WatsonxPassword?:string;
+  WatsonxProjectId?: string;
```
Comment (rishacha) [inline core/index.d.ts]: Follow camel-case please

File: core/llm/index.ts [hunk 3]
```diff
@@ -103,6 +103,16 @@ export abstract class BaseLLM implements ILLM {
...
+  // For WatsonX only.
+
+  WatsonxUrl?: string;
+  WatsonxApiKey?: string;
+  WatsonxZenApiKeyBase64?:string = "YOUR_WATSONX_ZENAPIKEY" // Required if using watsonx software with ZenApiKey auth
+  WatsonxUsername?:string;
+  WatsonxPassword?:string;
+  WatsonxProjectId?: string;
+
+
```
Comment (rishacha) [inline core/llm/index.ts]: Camel Case

File: core/llm/index.ts [hunk 4]
```diff
@@ -154,6 +164,13 @@ export abstract class BaseLLM implements ILLM {
     this.apiKey = options.apiKey;
     this.aiGatewaySlug = options.aiGatewaySlug;
     this.apiBase = options.apiBase;
+    // for watsonx only
+    this.WatsonxUrl = options.WatsonxUrl;
+    this.WatsonxApiKey = options.WatsonxApiKey;
+    this.WatsonxProjectId = options.WatsonxProjectId;
+    this.WatsonxUsername = options.WatsonxUsername;
+    this.WatsonxPassword = options.WatsonxPassword;
```
Comment (rishacha) [inline core/llm/index.ts]: Camel Case

File: gui/src/pages/AddNewModel/ConfigureProvider.tsx [hunk 5]
```diff
@@ -185,6 +212,42 @@ function ConfigureProvider() {
...
+                        required: false,
+                      })}
+                    />
+                  </div>
+                );
+              })}
+            <ContinueButton onClick={handleContinue}
+              disabled={disableModelCards() || (
+                modelInfo?.collectInputFor
+                  ?.filter((d) => d.iswatsonxattribute)
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/ConfigureProvider.tsx]: Camel Case

**Review by rishacha — COMMENTED** [review]
Camel Case fixes

File: core/llm/llms/WatsonX.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+              data.results.forEach((result: any) => {
+                generatedChunk += result.generated_text || "";
+              });
+            } catch (e) {
+              console.error(e);
+            }
+          }
+        });
+        generatedChunk = generatedChunk.replaceAll("<|im_end|>", " ");
+        generatedChunk = generatedChunk.replaceAll("<|im_start|> ", "\n");
```
Comment (rishacha) [inline core/llm/llms/WatsonX.ts:214]: Add conditions for Granite model

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 7]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
     ],
     collectInputFor: [...completionParamsInputsConfigs],
   },
+  watsonx: {
+    title: "WatsonX",
+    provider: "watsonx",
+    refPage: "watsonX",
+    description:
+      "Explore foundation models from IBM and other third-parties depending on your use case.",
+    longDescription: `WatsonX`,
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts]: Add  a better description

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 8]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
...
+    collectInputFor: [
+      {
+        inputType: "text",
+        key: "watsonxUrl",
+        label: "WatsonX URL",
+        placeholder: "Enter your WatsonxUrl",
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts]: Add example in the placeholder

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 9]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
...
+        required: true,
+      },
+      {
+        inputType: "text",
+        key: "watsonxApiKey",
+        label: "WatsonX API Key",
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts:532]: Add examples

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 10]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
...
+        inputType: "text",
+        key: "watsonxProjectId",
+        label: "WatsonX Project Id",
+        placeholder: "Enter your project Id",
+        required: true,
+      },
+      {
+        inputType: "text",
+        key: "watsonxUsername",
+        label: "WatsonX username",
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts]: Add examples

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 11]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
...
+        inputType: "text",
+        key: "watsonxApiKey",
+        label: "WatsonX API Key",
+        placeholder: "Enter your API key",
+        required: true,
+      },
+      {
+        inputType: "text",
+        key: "watsonxProjectId",
+        label: "WatsonX Project Id",
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts:540]: Add a sample

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 12]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
...
+        inputType: "text",
+        key: "title",
+        label: "Model name",
+        placeholder: "Enter the Model name",
+        isWatsonxAttribute: true,
+      },
+      {
+        inputType: "text",
+        key: "model",
+        label: "Model Id",
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts:572]: Keep granite as default

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 13]
```diff
@@ -483,4 +484,73 @@ After it's up and running, you can start using Continue.`,
...
+        key: "wsatsonxPassword",
+        label: "WatsonX Password",
+        placeholder: "Enter your password",
+        required: true,
+      },
+      {
+        inputType: "text",
+        key: "title",
+        label: "Model name",
+        placeholder: "Enter the Model name",
```
Comment (rishacha) [inline gui/src/pages/AddNewModel/configs/providers.ts]: Keep granite as default

**Comment by sestinj** [comment]
Look good! I think I'll pull this down and make a few adjustments, but overall excited to have first-class support for WatsonX models!

File: core/llm/llms/WatsonX.ts [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+            if (isNaN(generatedTextIndex)) { console.error(`Unable to parse stream chunk ID: ${el}`); }
+          } else if (el.startsWith("data:")) {
+            const dataStr = el.replace(/^data:\s+/, "");
+            try {
+              const data = JSON.parse(dataStr);
+              data.results.forEach((result: any) => {
+                generatedChunk += result.generated_text || "";
+              });
+            } catch (e) {
+              console.error(e);
```
Comment (rishacha) [inline core/llm/llms/WatsonX.ts]: Add proper error handling here

**Review by rishacha — COMMENTED** [review]
Make these changes

**Comment by rishacha** [comment]
@sestinj - Thank you for merging ! We were done and ready for a review 😃

**Pull request merged**